plugins/ffmpeg_plugin/kahawai.c

```diff
@@ -0,0 +1,606 @@
+  } else if (sscanf(s->src_addr, "%u.%u.%u.%u", &ops_rx.port.sip_addr[ST_PORT_P][0],
+                    &ops_rx.port.sip_addr[ST_PORT_P][1],
+                    &ops_rx.port.sip_addr[ST_PORT_P][2],
+                    &ops_rx.port.sip_addr[ST_PORT_P][3]) != ST_IP_ADDR_LEN) {
+    av_log(ctx, AV_LOG_ERROR, "Failed to parse source IP address: %s\n", s->src_addr);
+    return AVERROR(EINVAL);
+  }
+
+  if ((s->udp_port < 0) || (s->udp_port > 0xFFFF)) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid UDP port: %d\n", s->udp_port);
+    return AVERROR(EINVAL);
+  }
+  ops_rx.port.udp_port[ST_PORT_P] = s->udp_port;
+
+  if (s->width <= 0) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid transport width: %d\n", s->width);
+    return AVERROR(EINVAL);
+  }
+  ops_rx.width = s->width;
+
+  if (s->height <= 0) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid transport height: %d\n", s->height);
+    return AVERROR(EINVAL);
+  }
+  ops_rx.height = s->height;
+
+  if ((pix_fmt = av_get_pix_fmt(s->pixel_format)) == AV_PIX_FMT_NONE) {
+    av_log(ctx, AV_LOG_ERROR, "No such pixel format: %s.\n", s->pixel_format);
+    return AVERROR(EINVAL);
+  } else if (pix_fmt != AV_PIX_FMT_YUV422P10LE) {
+    av_log(ctx, AV_LOG_ERROR, "Only yuv422p10le is supported\n");
+    return AVERROR(EINVAL);
+  }
+  ops_rx.transport_fmt = ST20_FMT_YUV_422_10BIT;
+  ops_rx.output_fmt = ST_FRAME_FMT_YUV422RFC4175PG2BE10;
+
+  packet_size = av_image_get_buffer_size(pix_fmt, s->width, s->height, 1);
+  if (packet_size < 0) {
+    av_log(ctx, AV_LOG_ERROR, "av_image_get_buffer_size failed with %d\n", packet_size);
+    return packet_size;
+  }
+  av_log(ctx, AV_LOG_VERBOSE, "packet size: %d\n", packet_size);
+
+  fps = s->framerate.num * 100 / s->framerate.den;
+  for (ret = 0; ret < KAHAWAI_FPS_SUPPORTED; ++ret) {
+    if ((fps >= fps_table[ret][1]) && (fps <= fps_table[ret][2])) {
+      ops_rx.fps = fps_table[ret][0];
+      break;
+    }
+  }
+  if (ret >= KAHAWAI_FPS_SUPPORTED) {
+    av_log(ctx, AV_LOG_ERROR, "Frame rate %0.2f is not supported\n", ((float)fps / 100));
+    return AVERROR(EINVAL);
+  }
+
+  st = avformat_new_stream(ctx, NULL);
+  if (!st) {
+    return AVERROR(ENOMEM);
+  }
+
+  st->codecpar->codec_type = AVMEDIA_TYPE_VIDEO;
+  st->codecpar->codec_id = ctx->iformat->raw_codec_id;
+  st->codecpar->format = pix_fmt;
+  st->codecpar->width = s->width;
+  st->codecpar->height = s->height;
+  avpriv_set_pts_info(st, 64, s->framerate.den, s->framerate.num);
+  ctx->packet_size = packet_size;
+  st->codecpar->bit_rate =
+      av_rescale_q(ctx->packet_size, (AVRational){8, 1}, st->time_base);
+
+  // Create device
+  s->dev_handle = st_init(&param);
+  if (!s->dev_handle) {
+    av_log(ctx, AV_LOG_ERROR, "st_init failed\n");
+    return AVERROR(EIO);
+  }
+  av_log(ctx, AV_LOG_VERBOSE, "st_init finished\n");
+
+  ops_rx.name = "st20p";
+  ops_rx.priv = s;                 // Handle of priv_data registered to lib
+  ops_rx.port.payload_type = 112;  // RX_ST20_PAYLOAD_TYPE
```

Comment:
payload_type should be customized by command line, maybe we can enhance it later.

Reply:
Will enhance it later.